### PR TITLE
Water trimer descriptor

### PR DESCRIPTION
**Pull request by sujona**
Fast water trimer descriptor. Let me know what you think!

### CODE REVIEW

File: descriptors_noncommercial.inc [hunk 1]
```diff
@@ -3018,8 +3020,17 @@
                            !normalisation factor
                            descriptor_out%x(i_desc)%grad_covariance_cutoff(:,:) = descriptor_out%x(i_desc)%grad_covariance_cutoff(:,:) / n_products
                         end if cutoff_type_grad
-
+                        
                      end if calc_grad_descriptor
+
```
Comment (albapa) [inline descriptors_noncommercial.inc:3025]: Can you please remove this or change it to use our built in `print()` interface.

File: descriptors_noncommercial.inc [hunk 2]
```diff
@@ -4826,4 +4840,495 @@
...
+        ! If no Lattice is given: in quip the box is set to something random but big enough, in quippy to 0. 
+        if (all(box<1)) box = 1000 ! If box is 0 (<1) set it to very big
+        forall (i=1:3) lattice(i,i) = box(i)
+        
+        ok_ohh = all(at%Z(1::3) == 8) .and. all(at%Z(2::3) == 1) .and. all(at%Z(3::3) == 1)
+        ok_ort = at%is_orthorhombic
+        ok_box = all(box > 2*this%cutoff)
+        
+        if (.not.all([ok_ohh,ok_ort,ok_box]))then
+            STOP  __FILE__//":"//__LINE__//": water_trimer_calc: Inapplicable: ohh_order="//ok_ohh//" orthogonal="//ok_ort//" large_enough="//ok_box//" this box="//box
```
Comment (albapa) [inline descriptors_noncommercial.inc:5047]: can you please rewrite this with our predefined macros?

File: descriptors_noncommercial.inc [hunk 3]
```diff
@@ -4826,4 +4840,495 @@
...
+            ! RAISE_ERROR("water_trimer_calc: Inapplicable: ohh_order="//ok_ohh//" orthogonal="//ok_ort//" large_enough="//ok_box, error)
+        endif
+        
+        n_atom = at%N
+        n_mono = n_atom/3
+        ! imonomers = reshape([(i,i=1,n_atom)],[3,n_mono])
+        call find_triplets_lars(at%pos,n_mono,lattice,this%cutoff,monomer_triplets,triplets_diffs,triplets_diffs_map)
+        n_descriptors = size(triplets_diffs_map)
+        
+        ! print*, "  wt calc: found #mono="//n_mono//" #trip="//n_descriptors
```
Comment (albapa) [inline descriptors_noncommercial.inc:5057]: please remove these or re-write.

File: gap_fit_module.f95 [hunk 4]
```diff
@@ -741,19 +741,24 @@ subroutine read_descriptors(this)
     integer :: i
 
     this%max_cutoff = 0.0_dp
-
+    print*, "   read_descriptors 1"
```
Comment (albapa) [inline gap_fit_module.f95:744]: Please remove or re-write.

**Review by albapa — CHANGES_REQUESTED** [review]

File: descriptors_noncommercial.inc (hunk 1 above)
Comment (sujona) [reply descriptors_noncommercial.inc:3025]: Yes, that's a mistake, I'll remove all those prints.

**Review by sujona — COMMENTED** [review]
I have fixed @albapa's  comments

**Comment by albapa** [comment]
Interesting. I'd much rather have the macro but it should be functional. I might look at this later.

**Comment by albapa** [comment]
Could you please remove the `forall` statements? They're becoming deprecated.

**Comment by albapa** [comment]
Do all the tests run with your compiled code? Both serial, openmp and MPI (the latter for fitting)?

**Comment by sujona** [comment]
OK, I made a function that checks the lattice without forall. I fixed some other things like removing the shift_map that wasn't used. I compiled the MPI version and ran the tests, which came out fine (but 8 were skipped, and there is a test of calc_connect that allways fails (for me at least)) . I didn't manage to test the training with MPI myself becasue our MPI seems messed up for some (unrelated) reason. I've tested the water_trimer gradient's with finite difference.

**Comment by sujona** [comment]
I pushed another commit that repairs the general_trimer_calc (NOT water_trimer) gradients. They were wrong for power/=1. Maybe this wasn't the right time but hopefully it can all go through.

**Review by albapa — APPROVED** [review]

**Comment by albapa** [comment]
@sujona you'll need another pull request in QUIP that makes the GAP submodule point to the HEAD of this repo. It's in the README how to do it.

**Comment by sujona** [comment]
OK, it's done.

**Pull request merged**